File: api/app/services/digest_service.py
from datetime import date

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.digest import Digest
from app.schemas.digest import DigestCreate


async def create_digest(session: AsyncSession, data: DigestCreate) -> Digest:
    digest = Digest(**data.model_dump())
    session.add(digest)
    await session.flush()
    return digest


async def get_digests(
    session: AsyncSession,
    page: int = 1,
    per_page: int = 20,
) -> tuple[list[Digest], int]:
    count_result = await session.execute(select(func.count()).select_from(Digest))
    total = count_result.scalar_one()

    result = await session.execute(
        select(Digest)
        .order_by(Digest.digest_date.desc())
        .offset((page - 1) * per_page)
        .limit(per_page)
    )
    digests = list(result.scalars().all())
    return digests, total


async def get_digest_by_date(session: AsyncSession, digest_date: date) -> Digest | None:
    result = await session.execute(
        select(Digest).where(Digest.digest_date == digest_date)
    )
    return result.scalar_one_or_none()

File: api/src/services/digest-service.ts
import { count, desc, eq } from "drizzle-orm";
import type { PostgresJsDatabase } from "drizzle-orm/postgres-js";
import type * as schema from "../db/schema/index.js";
import { digests } from "../db/schema/index.js";
import type { DigestCreate } from "../schemas/digest.js";

type DB = PostgresJsDatabase<typeof schema>;

export async function createDigest(db: DB, data: DigestCreate) {
  const [digest] = await db
    .insert(digests)
    .values({
      digestDate: data.digest_date,
      title: data.title ?? null,
      content: data.content ?? null,
      articleCount: data.article_count ?? null,
      articleIds: data.article_ids ?? null,
    })
    .returning();
  return digest;
}

export async function getDigests(
  db: DB,
  page = 1,
  perPage = 20,
): Promise<{ items: (typeof digests.$inferSelect)[]; total: number }> {
  const [totalResult] = await db.select({ count: count() }).from(digests);
  const total = totalResult.count;

  const items = await db
    .select()
    .from(digests)
    .orderBy(desc(digests.digestDate))
    .offset((page - 1) * perPage)
    .limit(perPage);

  return { items, total };
}

export async function getDigestByDate(db: DB, digestDate: string) {
  const [digest] = await db.select().from(digests).where(eq(digests.digestDate, digestDate));
  return digest ?? null;
}
